toolkits/slack/arcade_slack/utils.py

```diff
@@ -0,0 +1,502 @@
+import asyncio
+from datetime import datetime, timezone
+from typing import Callable, Literal, NewType, NotRequired, Optional, TypedDict
+
+from arcade.sdk.errors import RetryableToolError
+
+from arcade_slack.constants import MAX_PAGINATION_SIZE_LIMIT, MAX_PAGINATION_TIMEOUT_SECONDS
+from arcade_slack.models import ConversationType, ConversationTypeSlackName
+from arcade_slack.tools.exceptions import PaginationTimeoutError
+
+SlackOffsetSecondsFromUTC = NewType("SlackOffsetSecondsFromUTC", int)  # observe it can be negative
+SlackPaginationNextCursor = NewType("SlackPaginationNextCursor", str)
+SlackUserFieldId = NewType("SlackUserFieldId", str)
+SlackUserId = NewType("SlackUserId", str)
+SlackTeamId = NewType("SlackTeamId", str)
+SlackTimestampStr = NewType("SlackTimestampStr", str)
+
+"""
+About Slack types in general: Slack does not guarantee the presence of all fields for a given
+object. It will vary from endpoint to endpoint and even if the field is present, they say it may
+contain a None value or an empty string instead of the actual expected value.
+
+Because of that, our TypedDicts ended up having to be mostly total=False and most of the fields'
+type hints are Optional. Use Slack dictionary fields with caution. It's advisable to validate the
+value before using it, so that we can raise errors that are clear to understand.
+
+See, for example, the 'Common Fields' section of the user type definition at:
+https://api.slack.com/types/user#fields (https://archive.is/RUZdL)
+"""
+
+
+class SlackUserFieldData(TypedDict, total=False):
```

Comment:
@Spartee please take a look at the dict typing to see if it's in line with what you have in mind